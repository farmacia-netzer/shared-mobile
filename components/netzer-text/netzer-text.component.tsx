import { FONT_SIZE } from '##theme/typography.constant';
import React, { ReactNode, useCallback, useMemo } from 'react';
import { StyleProp, StyleSheet, Text, TextProps, TextStyle } from 'react-native';
import { brandStyles, coreStyles, customStyles } from './netzer-text.style';

type TBrandStyles = typeof brandStyles;
type TCoreStyles = typeof coreStyles;
type TCustomStyles = typeof customStyles;
type TStyleNames = TBrandStyles & TCoreStyles & TCustomStyles;

interface NetzerTextProps extends TextProps {
  text: string | ReactNode;
  styleName?: keyof TStyleNames;
  textColor?: any;
  ellipsizeAtLine?: number;
  fontStyle?: any;
  type?: TTitleSizes;
}

enum ETitleSizes {
  TITLE = 'TITLE',
  SUBTITLE = 'SUBTITLE',
  TEXT = 'TEXT'
}

type TTitleSizes = 'TITLE' | 'SUBTITLE' | 'TEXT';

export const NetzerText: React.FC<NetzerTextProps> = ({
  text,
  styleName = 'NORMAL',
  ellipsizeAtLine = 0,
  type = 'TEXT',
  ...rest
}: NetzerTextProps) => {
  const { style, ...otherTextProps } = rest;

  const textStyle: TextStyle =
    coreStyles[styleName] || customStyles[styleName] || brandStyles[styleName] || coreStyles.NORMAL;

  // const fontFamily = getFontFamily(styleName, fontStyle);

  // const styleProps: StyleProp<TextStyle> = useMemo(
  //   () => [textStyle, fontFamily ? { fontFamily } : undefined, style],
  //   [fontFamily, style, textStyle]
  // );

  const getTitleSize = useCallback(() => {
    return {
      [ETitleSizes.TITLE]: styles.title,
      [ETitleSizes.SUBTITLE]: styles.subtitle,
      [ETitleSizes.TEXT]: styles.text
    }?.[type];
  }, [type]);

  const styleProps: StyleProp<TextStyle> = useMemo(
    () => [textStyle, undefined, style, getTitleSize()],
    [getTitleSize, style, textStyle]
  );

  const props = {
    ...(ellipsizeAtLine > 0 && {
      numberOfLines: ellipsizeAtLine
    }),
    ...otherTextProps
  };

  return (
    <Text style={styleProps} {...props}>
      {text}
    </Text>
  );
};

const styles = StyleSheet.create({
  title: {
    fontSize: FONT_SIZE.LARGE,
    fontWeight: 'bold',
    fontFamily: 'Avenir Next',
    paddingVertical: 3
  },
  subtitle: {
    fontSize: FONT_SIZE.NORMAL
  },
  text: {}
});
