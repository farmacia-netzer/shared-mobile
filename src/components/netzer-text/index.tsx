export * from './netzer-text.component';
export * from './netzer-text.style';
