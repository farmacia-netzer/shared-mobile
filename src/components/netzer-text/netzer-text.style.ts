import { TEXT_COLORS } from '../../theme/colors.constant';
import { SPECIAL_FONT_SIZE, LINE_HEIGHT, FONT_SIZE } from '../../theme/typography.constant';
import { StyleSheet } from 'react-native';


const lineHeightAspectRatio = 1.3;

// export const getFontFamily = (styleName, fontStyle) => {
//   if (brandStyles[styleName] || customStyles[styleName]) {
//     return null;
//   }
//   return CORE_FONT[fontStyle] || CORE_FONT.REGULAR;
// };

export const brandStyles = StyleSheet.create({
  BRAND_XXLARGE: {
    // fontFamily: theme.brandFont.family.MEDIUM,
    fontSize: SPECIAL_FONT_SIZE.BRAND_XXLARGE,
    lineHeight: LINE_HEIGHT.XXLARGE
    // ...theme.brandFont.style.XXLARGE
  },
  BRAND_XLARGE: {
    // fontFamily: theme.brandFont.family.LIGHT,
    fontSize: SPECIAL_FONT_SIZE.BRAND_XLARGE,
    lineHeight: LINE_HEIGHT.XLARGE
    // ...theme.brandFont.style.XLARGE
  },
  BRAND_MEDIUM: {
    // fontFamily: theme.brandFont.family.LIGHT,
    fontSize: FONT_SIZE.MEDIUM,
    lineHeight: LINE_HEIGHT.MEDIUM
    // ...theme.brandFont.style.MEDIUM
  },
  BRAND_MEDIUM_BOLD: {
    // fontFamily: theme.brandFont.family.BOLD,
    fontSize: FONT_SIZE.MEDIUM,
    lineHeight: LINE_HEIGHT.MEDIUM
    // ...theme.brandFont.style.MEDIUM_BOLD
  },
  // Used only for header page title
  BRAND_HEADER_TITLE: {
    color: TEXT_COLORS.PRIMARY,
    // fontFamily: theme.brandFont.family.BOLD,
    fontSize: FONT_SIZE.MEDIUM
    // ...theme.brandFont.style.HEADER_TITLE
  }
});

export const coreStyles = StyleSheet.create({
  XXSMALL: {
    marginVertical: -2,
    fontSize: FONT_SIZE.XXSMALL,
    lineHeight: FONT_SIZE.XXSMALL + 4
  },
  XSMALL: {
    fontSize: FONT_SIZE.XSMALL,
    lineHeight: LINE_HEIGHT.XSMALL + 4
  },
  XSMALL_MULTILINE: {
    fontSize: FONT_SIZE.XSMALL,
    lineHeight: LINE_HEIGHT.XSMALL * lineHeightAspectRatio
  },
  SMALL: {
    fontSize: FONT_SIZE.SMALL,
    lineHeight: FONT_SIZE.SMALL + 4
  },
  SMALL_MULTILINE: {
    fontSize: FONT_SIZE.SMALL,
    lineHeight: LINE_HEIGHT.SMALL
  },
  SUBTEXT: {
    fontSize: FONT_SIZE.SUBTEXT,
    lineHeight: FONT_SIZE.SUBTEXT + 4
  },
  SUBTEXT_MULTILINE: {
    fontSize: FONT_SIZE.SUBTEXT,
    lineHeight: LINE_HEIGHT.SUBTEXT
  },
  NORMAL: {
    fontSize: FONT_SIZE.NORMAL,
    lineHeight: FONT_SIZE.NORMAL + 4
  },
  NORMAL_MULTILINE: {
    fontSize: FONT_SIZE.NORMAL,
    lineHeight: LINE_HEIGHT.NORMAL
  },
  MEDIUM: {
    fontSize: FONT_SIZE.MEDIUM,
    lineHeight: FONT_SIZE.MEDIUM + 4
  },
  MEDIUM_MULTILINE: {
    fontSize: FONT_SIZE.MEDIUM,
    lineHeight: LINE_HEIGHT.MEDIUM
  },
  LARGE: {
    fontSize: FONT_SIZE.LARGE,
    lineHeight: LINE_HEIGHT.LARGE
  },
  LARGE_MULTILINE: {
    fontSize: FONT_SIZE.LARGE,
    lineHeight: LINE_HEIGHT.LARGE_MULTILINE
  },
  XLARGE: {
    fontSize: FONT_SIZE.XLARGE,
    lineHeight: LINE_HEIGHT.XLARGE
  },
  XXLARGE: {
    fontSize: FONT_SIZE.XXLARGE,
    lineHeight: LINE_HEIGHT.XXLARGE
  }
});

export const customStyles = StyleSheet.create({
  ERROR: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: 16,
    color: TEXT_COLORS.NEGATIVE
  },
  INPUT: {
    color: TEXT_COLORS.CONTRAST_HIGH,
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.NORMAL
  },
  INPUT_ERROR: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.SMALL,
    lineHeight: 16,
    color: TEXT_COLORS.NEGATIVE
  },
  SECTION_HEADER: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.MEDIUM,
    textTransform: 'uppercase',
    color: TEXT_COLORS.CONTRAST_MIN
  },
  LIST_SEPARATOR: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.SMALL,
    textTransform: 'uppercase',
    color: TEXT_COLORS.CONTRAST_MIN
  },
  UPPERCASE_DATE: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.SMALL,
    textTransform: 'uppercase',
    color: TEXT_COLORS.CONTRAST_LOW
  },
  ICON_CAPTION: {
    color: TEXT_COLORS.CONTRAST_LOW,
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.SUBTEXT,
    lineHeight: FONT_SIZE.SUBTEXT + 4
  },
  TAB_BAR_ICON_CAPTION: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.XSMALL,
    letterSpacing: -0.2,
    lineHeight: 13,
    textTransform: 'uppercase'
  },
  SMALL_BUTTON_LIST_TITLE_TEXT: {
    // fontFamily: CORE_FONT.REGULAR,
    fontSize: FONT_SIZE.XSMALL,
    lineHeight: 10 * lineHeightAspectRatio,
    textTransform: 'uppercase',
    color: TEXT_COLORS.CONTRAST_LOW
  },
  CALORIES_TITLE: {
    // fontFamily: CORE_FONT.EXTRABOLD,
    fontSize: SPECIAL_FONT_SIZE.CALORIES_TITLE,
    lineHeight: SPECIAL_FONT_SIZE.CALORIES_TITLE + 4
  },
  CALORIES_VALUE: {
    // fontFamily: CORE_FONT.EXTRABOLD,
    fontSize: SPECIAL_FONT_SIZE.CALORIES_VALUE,
    lineHeight: SPECIAL_FONT_SIZE.CALORIES_VALUE + 4
  }
});
