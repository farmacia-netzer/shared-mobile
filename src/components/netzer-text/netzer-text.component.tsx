
import React, { ReactNode, useCallback, useMemo } from 'react';
import { StyleProp, StyleSheet, Text, TextProps, TextStyle } from 'react-native';
import { brandStyles, coreStyles, customStyles } from './netzer-text.style';
import { SharedElement } from 'react-navigation-shared-element';
import { FONT_SIZE } from '../../theme/typography.constant';

type TBrandStyles = typeof brandStyles;
type TCoreStyles = typeof coreStyles;
type TCustomStyles = typeof customStyles;
type TStyleNames = TBrandStyles & TCoreStyles & TCustomStyles;

interface NetzerTextProps extends TextProps {
  text: string | ReactNode;
  styleName?: keyof TStyleNames;
  textColor?: any;
  ellipsizeAtLine?: number;
  fontStyle?: any;
  type?: TTitleSizes;
  sharedElement?: boolean;
  sharedElementId?: string;
}

enum ETitleSizes {
  TITLE = 'TITLE',
  SUBTITLE = 'SUBTITLE',
  TEXT = 'TEXT'
}

type TTitleSizes = 'TITLE' | 'SUBTITLE' | 'TEXT';

export const NetzerText: React.FC<NetzerTextProps> = ({
  text,
  styleName = 'NORMAL',
  ellipsizeAtLine = 0,
  type = 'TEXT',
  sharedElement,
  sharedElementId,
  ...rest
}: NetzerTextProps) => {
  const { style, ...otherTextProps } = rest;

  const textStyle: TextStyle =
    coreStyles[styleName] || customStyles[styleName] || brandStyles[styleName] || coreStyles.NORMAL;

  // const fontFamily = getFontFamily(styleName, fontStyle);

  // const styleProps: StyleProp<TextStyle> = useMemo(
  //   () => [textStyle, fontFamily ? { fontFamily } : undefined, style],
  //   [fontFamily, style, textStyle]
  // );

  const getTitleSize = useCallback(() => {
    return {
      [ETitleSizes.TITLE]: styles.title,
      [ETitleSizes.SUBTITLE]: styles.subtitle,
      [ETitleSizes.TEXT]: styles.text
    }?.[type];
  }, [type]);

  const styleProps: StyleProp<TextStyle> = useMemo(
    () => [textStyle, undefined, style, getTitleSize()],
    [getTitleSize, style, textStyle]
  );

  const props = {
    ...(ellipsizeAtLine > 0 && {
      numberOfLines: ellipsizeAtLine
    }),
    ...otherTextProps
  };

  const TextComponent = (
    <Text style={styleProps} {...props}>
      {text}
    </Text>
  );

  const Content = sharedElement ? <SharedElement id={sharedElementId!}>{TextComponent}</SharedElement> : TextComponent;
  return Content;
};

const textStyles = {
  fontFamily: 'Avenir Next'
}
const styles = StyleSheet.create({
  title: {
    ...textStyles,
    fontSize: FONT_SIZE.LARGE,
    fontWeight: 'bold',
    fontFamily: 'Avenir Next',
    paddingVertical: 3
  },
  subtitle: {
    ...textStyles,
    fontSize: FONT_SIZE.NORMAL
  },
  text: {
    ...textStyles
  }
});
