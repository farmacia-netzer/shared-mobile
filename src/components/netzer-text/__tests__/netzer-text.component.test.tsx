import React from 'react';
import { render } from '@testing-library/react-native';
import { NetzerText } from '../netzer-text.component';

describe('<NetzerText />', () => {
  const testText = 'test text';

  it('renders correctly', () => {
    const { toJSON } = render(<NetzerText text={testText} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
