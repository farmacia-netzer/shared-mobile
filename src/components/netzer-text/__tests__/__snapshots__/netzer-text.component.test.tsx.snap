// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NetzerText /> renders correctly 1`] = `
<Text
  style={
    Array [
      Object {
        "fontSize": 14,
        "lineHeight": 18,
      },
      undefined,
      undefined,
    ]
  }
>
  test text
</Text>
`;
